### ISSUE TITLE
data loss on windows platform

### ISSUE THREAD

**Issue by michael2008** [linkedin/ambry#594]
Originally, we are using Ambry on linux and everything is fine. But recently, we try to use Ambry on the client side where the OS is Windows 10. After we restart Ambry, all data lost. Log says:

```
[2017-04-13 09:06:12,176] INFO Bootstrapping AmbryServer (com.github.ambry.server.AmbryMain)
[2017-04-13 09:06:12,179] INFO starting (com.github.ambry.server.AmbryServer)
[2017-04-13 09:06:12,179] INFO Setting up JMX. (com.github.ambry.server.AmbryServer)
[2017-04-13 09:06:12,240] INFO creating configs (com.github.ambry.server.AmbryServer)
[2017-04-13 09:06:12,243] INFO Verifying properties (com.github.ambry.config.VerifiableProperties)
[2017-04-13 09:06:12,244] INFO Property host.name is overridden to localhost (com.github.ambry.config.VerifiableProperties)
[2017-04-13 09:06:12,244] INFO Property store.data.flush.interval.seconds is overridden to 3 (com.github.ambry.config.VerifiableProperties)
[2017-04-13 09:06:12,245] INFO check if node exist in clustermap host localhost port 6667 (com.github.ambry.server.AmbryServer)
[2017-04-13 09:06:12,258] INFO Starting storage manager (com.github.ambry.store.StorageManager)
[2017-04-13 09:06:12,263] INFO Loading segment with name [] (com.github.ambry.store.Log)
[2017-04-13 09:06:12,264] INFO Segment [] has capacity of 10737418240 bytes (com.github.ambry.store.Log)
[2017-04-13 09:06:12,487] INFO Index :  C:\Users\DM\Desktop\oss\data\0 log end offset of index  before recovery Name = [] Offset = [0] (com.github.ambry.store.PersistentIndex)
[2017-04-13 09:06:12,487] INFO Index :  C:\Users\DM\Desktop\oss\data\0 performing recovery on index with start offset Name = [] Offset = [0] and end offset 506 (com.github.ambry.store.PersistentIndex)
[2017-04-13 09:06:12,487] ERROR Trying to read more than the available bytes (com.github.ambry.messageformat.BlobStoreRecovery)
[2017-04-13 09:06:12,489] INFO Setting active segment to [] (com.github.ambry.store.Log)
[2017-04-13 09:06:12,489] INFO Index :  C:\Users\DM\Desktop\oss\data\0 Starting hard delete recovery (com.github.ambry.store.PersistentIndex)
[2017-04-13 09:06:12,489] INFO Index : C:\Users\DM\Desktop\oss\data\0 Finished performing hard delete recovery (com.github.ambry.store.PersistentIndex)
[2017-04-13 09:06:12,492] INFO Starting storage manager complete (com.github.ambry.store.StorageManager)
[2017-04-13 09:06:12,493] INFO BlockingChannelConnectionPool started (com.github.ambry.network.BlockingChannelConnectionPool)
[2017-04-13 09:06:12,500] INFO Reading replica tokens for mount path C:\Users\DM\Desktop\oss\data (com.github.ambry.replication.ReplicationManager)
[2017-04-13 09:06:12,500] WARN Number of Datacenters to replicate from is 0, not starting any replica threads (com.github.ambry.replication.ReplicationManager)
[2017-04-13 09:06:12,510] INFO Starting 8 processor threads (com.github.ambry.network.SocketServer)
[2017-04-13 09:06:12,561] INFO Starting acceptor threads (com.github.ambry.network.SocketServer)
[2017-04-13 09:06:12,562] INFO Awaiting socket connections on localhost:6667 (com.github.ambry.network.Acceptor)
[2017-04-13 09:06:12,563] INFO Started server (com.github.ambry.network.SocketServer)
[2017-04-13 09:06:12,563] INFO started (com.github.ambry.server.AmbryServer)
[2017-04-13 09:06:12,563] INFO Server startup time in Ms 383 (com.github.ambry.server.AmbryServer)
[2017-04-13 09:06:59,547] INFO Creating first segment (com.github.ambry.store.PersistentIndex)
```

I am aware of that Ambry is not tested on Windows. But according to the error log "Trying to read more than the available bytes", this might be fixable. Will you guys fix it on Windows?

**Comment by michael2008** [comment]
Here is the test process on Windows (On linux, no problem):

After clean all the data in the data store, I post some data to Ambry:

```
curl -i -H "x-ambry-blob-size : `wc -c test.txt | xargs | cut -d" " -f1`" -H "x-ambry-service-id : CUrlUpload"  -H "x-ambry-owner-id : `whoami`" -H "x-ambry-content-type : image/gif" -H "x-ambry-um-description : Demonstration Image" http://192.168.1.45:1174/ --data-binary @test.txt                     HTTP/1.1 201 Created
Location: /AAEAAQAAAAAAAAAAAAAAJDJjZTMxNjRjLTEzYzMtNGZmYy1iYjM1LTU0MjcxOTdiMmU0OQ
Date: Thu, 13 Apr 2017 01:22:54 GMT
Content-Length: 0
x-ambry-creation-time: Thu, 13 Apr 2017 01:22:54 GMT
```

a few seconds later, I kill Ambry jvm via a .bat: 

```
FOR /F "tokens=5 delims= " %%P IN ('netstat -a -n -o ^| findstr :1174.*LISTENING') DO TaskKill.exe /F /PID %%P
FOR /F "tokens=5 delims= " %%P IN ('netstat -a -n -o ^| findstr :6667.*LISTENING') DO TaskKill.exe /F /PID %%P
```

then restart Ambry, the error log says what I posted before. In the datastore directory, "0_index.tmp" and "log_current" files both have data. But ambry failed to recover data from the restart.

**Comment by vgkholla** [comment]
@michael2008 "Trying to read more than the available bytes" is expected during recovery. It is the way the recovery code detects that nothing more can be recovered.

In your second example, when you restarted ambry did you see 0_index (not .tmp)? Is the error log you pasted above from this very restart? How big is the file you uploaded? ambry's flushing to disk is async (file system is also flushing regularly but we don't rely on that) - so it is possible that you lose some data if you kill the process. Recovery can only recover complete blobs. It stops when it encounters incomplete blobs. So you might have the killed the process here before all the data was written into the log.

**Comment by michael2008** [comment]
@vgkholla   actually, there is no .tmp file when I restart the server. No matter how long I wait, the recovery would fail. async should work, but it does not, don't know why.

**Comment by vgkholla** [comment]
@michael2008 @nsivabalan found the bug that is causing this. Will be fixed shortly. However, keep in mind that the log message can print during recovery even after the bug is fixed if a blob has been half written into the log at the event of a crash.

**Comment by vgkholla** [comment]
Being fixed in #602
